File: app/src/main/java/com/android/ashwiask/tvmaze/home/ShowDiffUtilCallback.java
package com.android.ashwiask.tvmaze.home;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;

import java.util.List;

public class ShowDiffUtilCallback extends DiffUtil.Callback {
    private final List<Show> oldList;
    private final List<Show> newList;

    public ShowDiffUtilCallback(List<Show> oldList, List<Show> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Show oldShow = oldList.get(oldItemPosition);
        Show newShow = newList.get(newItemPosition);
        return oldShow.id() == newShow.id();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Show oldShow = oldList.get(oldItemPosition);
        Show newShow = newList.get(newItemPosition);
        return oldShow.isFavorite() == newShow.isFavorite();
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        Show oldShow = oldList.get(oldItemPosition);
        Show newShow = newList.get(newItemPosition);
        Bundle bundle = new Bundle();
        if (oldShow.isFavorite() != newShow.isFavorite()) {
            bundle.putBoolean("IS_FAVORITE", newShow.isFavorite());
        }
        return bundle;
    }
}

File: app/src/main/java/com/android/ashwiask/tvmaze/home/ShowDiffUtilCallback.kt
package com.android.ashwiask.tvmaze.home

import android.os.Bundle
import androidx.recyclerview.widget.DiffUtil

class ShowDiffUtilCallback(
    private val oldList: List<HomeViewData.EpisodeViewData>,
    private val newList: List<HomeViewData.EpisodeViewData>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldEpisodeViewData = oldList[oldItemPosition]
        val newEpisodeViewData = newList[newItemPosition]
        return oldEpisodeViewData.showViewData.show.id == newEpisodeViewData.showViewData.show.id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldEpisodeViewData = oldList[oldItemPosition]
        val newEpisodeViewData = newList[newItemPosition]
        return oldEpisodeViewData.showViewData.isFavoriteShow == newEpisodeViewData.showViewData.isFavoriteShow
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        val oldEpisodeViewData = oldList[oldItemPosition]
        val newEpisodeViewData = newList[newItemPosition]
        val bundle = Bundle()
        if (oldEpisodeViewData.showViewData.isFavoriteShow != newEpisodeViewData.showViewData.isFavoriteShow) {
            bundle.putBoolean(IS_FAVORITE, newEpisodeViewData.showViewData.isFavoriteShow)
        }
        return bundle
    }
}
